subprojects/test-kit/src/integTest/groovy/org/gradle/testkit/runner/GradleRunnerCrossGroovyVersionIntegrationTest.groovy

```diff
@@ -0,0 +1,140 @@
+ * limitations under the License.
+ */
+
+package org.gradle.testkit.runner
+
+import org.gradle.integtests.fixtures.executer.GradleContextualExecuter
+import org.gradle.util.GradleVersion
+import spock.lang.IgnoreIf
+
+@IgnoreIf({ GradleContextualExecuter.embedded })
+class GradleRunnerCrossGroovyVersionIntegrationTest extends BaseGradleRunnerIntegrationTest {
+
+    def "can run simple build using groovy2"(String testedGradleVersion, Closure<GradleRunner> gradleVersionConfigurator) {
+        given:
+        buildFile << """
+            task writeGroovyVersion {
+                doLast {
+                    file("version.txt").with {
+                        createNewFile()
+                        text = GroovySystem.version
+                    }
+                }
+            }
+        """
+
+        when:
+        def someRunner = runner().withArguments("writeGroovyVersion")
+        gradleVersionConfigurator.call(someRunner)
+        someRunner.build()
+
+        then:
+        file("version.txt").text == groovyVersion
+
+        cleanup:
+        testKitDaemons(GradleVersion.version(testedGradleVersion)).killAll()
+
+        where:
+        testedGradleVersion               | groovyVersion | gradleVersionConfigurator
+        "6.8.3"                           |  "2.5.12"     |  { it.withGradleVersion("6.8.3") }
+        GradleVersion.current().version   | "3.0.7"       |  { if (!GradleContextualExecuter.embedded) { it.withGradleInstallation(buildContext.gradleHomeDir) } }
+    }
+
+    def "groovy2-based gradle can run build via testkit using a groovy3-based gradle"() {
+        given:
+        def targetingGradle7Test = """
+import spock.lang.Specification
+import org.junit.Rule
+import org.junit.rules.TemporaryFolder
+import org.gradle.testkit.runner.GradleRunner
+import org.gradle.testkit.runner.TaskOutcome
+import org.gradle.util.GradleVersion
+
+class BuildLogicFunctionalTest extends Specification {
+
+    @Rule public final TemporaryFolder testProjectDir = new TemporaryFolder();
+    File settingsFile
+    File buildFile
+
+    def setup() {
+        settingsFile = new File(testProjectDir.getRoot(), 'settings.gradle')
+        buildFile = new File(testProjectDir.getRoot(), 'build.gradle')
+    }
+
+    def "capture groovy version"() {
+        given:
+        settingsFile << "rootProject.name = 'hello-world'"
+        buildFile << \"\"\"
+            task writeGroovyVersion {
+                doLast {
+                    file("version.txt").with {
+                        createNewFile()
+                        text = GroovySystem.version
+                    }
+                }
+            }
+        \"\"\"
+
+        when:
+        def result = GradleRunner.create()
+            .withProjectDir(testProjectDir.getRoot())
+            .withArguments('writeGroovyVersion', '--stacktrace', '--info')
+            .withGradleInstallation(new File("${buildContext.gradleHomeDir}"))
```

Comment:
Would be interesting to test both `withDebug(true)` and `withDebug(false)` cases here.

Reply:
What does `withDebug` in particular to that makes you think will affect the test in questions?

Reply:
It runs the build embedded in the testkit process, without using a daemon at all.